spring-integration-core/src/main/java/org/springframework/integration/handler/support/MessagingMethodInvokerHelper.java

```diff
@@ -295,7 +296,14 @@ public void setUseSpelInvoker(boolean useSpelInvoker) {
 	@Override
 	public void setBeanFactory(@NonNull BeanFactory beanFactory) {
 		super.setBeanFactory(beanFactory);
-		this.messageHandlerMethodFactory.setBeanFactory(beanFactory);
+		if (isProvidedMessageHandlerFactoryBean()) {
+			this.messageHandlerMethodFactory = beanFactory.getBean(IntegrationContextUtils.MESSAGE_HANDLER_FACTORY_BEAN_NAME,
+					MessageHandlerMethodFactory.class);
+		}
```

Comment:
I don't think we can do this here; the beanFactory might not have the bean yet. 

Move to initialize?

Reply:
I thought about it and was struggling as well, but i guess you're right.
In all fairness I think this class needs a bit more surgery. Look at line 262 where `messageHandlerMethodFactory` is used which means we will createInvocableHandlerMethod *always* using the default one, yet may still inject one for the actual invocation. But I stop short of trying to make a big change. Perhaps "Consider revisiting..." type issue?

Reply:
Definitely better to move this `getBean()` into the `initialize()` below. This is not a "one-stop-shop" bean and it really may have a lot of dependencies, so calling it from the `setBeanFactory()` is a potential bug.

Reply:
I'll move it but keep in mind that it is still broken given what I said above